### ISSUE TITLE
[BUG] Vert.x "response has already been written" should be prevented

### ISSUE THREAD

**Issue by michael72** [softwaremill/tapir#3157]
Tapir version: v1.7.3

Scala version: 2.13.11

I use Tapir with vert.x and Future serving a remote webserver that gets some huge json file via PUT and a slow connection - all works fine when the file is smaller.
Due to reasons I haven't yet quite fathomed there is a `io.netty.channel.StacklessClosedChannelException: null` happening which itself leads to 

```log
java.lang.IllegalStateException: Response has already been written
at io.vertx.core.http.impl.Http1xServerResponse.end(Http1xServerResponse.java:421) ~[availability-simulator.jar:0.1.0-SNAPSHOT]
at io.vertx.core.http.impl.Http1xServerResponse.end(Http1xServerResponse.java:409) ~[availability-simulator.jar:0.1.0-SNAPSHOT]
at io.vertx.core.http.impl.Http1xServerResponse.end(Http1xServerResponse.java:478) ~[availability-simulator.jar:0.1.0-SNAPSHOT]
at sttp.tapir.server.vertx.VertxFutureServerInterpreter.$anonfun$endpointHandler$5(VertxFutureServerInterpreter.scala:73)
```

looking at the [code](https://github.com/softwaremill/tapir/blob/v1.7.3/server/vertx-server/src/main/scala/sttp/tapir/server/vertx/VertxFutureServerInterpreter.scala#L73) 
```scala
.foreach { ex =>
  logger.error("Error while processing the request", ex)
  if (rc.response().bytesWritten() > 0) rc.response().end()
    rc.fail(ex)
  }
```
it seems like the same problem happens as in [this stackoverflow request](https://stackoverflow.com/questions/59935129/vert-x-response-has-already-been-written) which according to this [answer](https://stackoverflow.com/a/59935408/2056545) could be solved as by checking the request has not `ended` before issuing the `end`:
```java
if (!req.response().ended()) {
    req.response().end(response.body());
}
```

i.e. this should probably fix the needless "response has already been written" error.

```scala
.foreach { ex =>
  logger.error("Error while processing the request", ex)
  if (rc.response().bytesWritten() > 0 && !req.response().ended()) rc.response().end()
    rc.fail(ex)
  }
```

=> the code lines were copy&pasted so the same problem should occur with vert.x and Cats + Zio

**Comment by adamw** [comment]
Do you know what is the response that is being written? Maybe some vertx-global timeout kicks in and closes the connection, or is it connection closed on the client side?

Either way, since you made a detail investigation, maybe you would be open to opening a PR with a fix?

**Comment by michael72** [comment]
Hello @adamw,

since the error logs occur only as an "afterthought" and all internal operations were successful and the response (accept) has also been written it is only the logs that are a bit annoying here. And of course it would be nice to find out what happens. It could either be an internal timeout or the amount of data being processed. I tried with a better internet connection and transfer rate and same error logs happen again. When I shorten the json file to 2/3rd of its original size it works fine.

OK - it is probably worth mentioning that - because the data file being processed is quite large and we have to save memory - I went from `jsonBody[List[Item]]` to `inputStreamBody` in the API - the generated documentation itself still uses the `jsonBody` endpoint. I also tried going via `fileBody` - same errors. I tried closing the stream manually - same errors. 
With the `jsonBody` itself or using `stringBody` (but then having the issue of needing a lot more memory) it all works fine - no error logs.
So I guess it could be a problem of how I use the API or maybe some internal configuration issue - either inside tapir or in vertx.

To the actual bugfix: I could add the code as mentioned - not sure (yet) about the code duplication though ;-) - and I would need a couple of days right now to find the time for it.

**Comment by michael72** [comment]
Hm - one question about the code again - I think I don't get it
[here](https://github.com/softwaremill/tapir/blob/v1.7.3/server/vertx-server/cats/src/main/scala/sttp/tapir/server/vertx/cats/VertxCatsServerInterpreter.scala#L74), [here](https://github.com/softwaremill/tapir/blob/v1.7.3/server/vertx-server/src/main/scala/sttp/tapir/server/vertx/VertxFutureServerInterpreter.scala#L73) or [here](https://github.com/softwaremill/tapir/blob/v1.7.3/server/vertx-server/zio/src/main/scala/sttp/tapir/server/vertx/zio/VertxZioServerInterpreter.scala#L51)

the [`end()` function](https://github.com/eclipse-vertx/vert.x/blob/master/src/main/java/io/vertx/core/http/HttpServerResponse.java#L277) itself returns a `Future[Void]` - is that future simply ignored in the code? Same goes for the `fail` code afterwards?

**Comment by adamw** [comment]
> To the actual bugfix: I could add the code as mentioned - not sure (yet) about the code duplication though ;-) - and I would need a couple of days right now to find the time for it.

No problem :) Yeah there is some code duplication, but that's the price you pay for integrating with 3 different runtimes.

> the [end() function](https://github.com/eclipse-vertx/vert.x/blob/master/src/main/java/io/vertx/core/http/HttpServerResponse.java#L277) itself returns a Future[Void] - is that future simply ignored in the code? Same goes for the fail code afterwards?

It seems so, I guess we're not interested at that point when the request will actually finish writing? But there might be sth wrong with that of course

**Comment by michael72** [comment]
OK - the bug with the doubled error message is definitely fixed in 1.8.0. I also tested with our product, where the error was reproducible before.

Thanks so much!
